File: ExploreHarness/Benchmarks/analyze.py
#!/usr/bin/env python3
"""Paired A/B analyzer for ExploreBenchmark JSONL output.

Usage:
    python3 Benchmarks/analyze.py baseline.jsonl candidate.jsonl \
        [--metrics attemptsPerSecond,coveredEdges,clusterCount,reducedTotal] \
        [--discovery "label=channel: 3;flags: 35"] ...

Records are paired by (fixture, seed); every comparison is per-seed first, aggregates second.
For each metric the script prints the per-seed table, medians and IQRs for both arms, the
median delta, and a two-sided paired sign-test verdict. Ties are dropped from the sign test.

--discovery defines an attempts-to-discovery metric: the minimum firstSeenAttempt over
clusters whose canonicalDescription contains every semicolon-separated marker. A run that
never found a matching cluster is censored: it counts as strictly worse than any run that
did, and censored-in-both pairs are dropped.
"""

import argparse
import json
import math
import sys
from statistics import median

DEFAULT_METRICS = ["attemptsPerSecond", "coveredEdges", "clusterCount", "reducedTotal"]


def load(path):
    records = {}
    with open(path) as handle:
        for line in handle:
            line = line.strip()
            if not line:
                continue
            record = json.loads(line)
            key = (record["fixture"], record["seed"])
            if key in records:
                print(f"warning: duplicate record for {key} in {path}; keeping the last one", file=sys.stderr)
            records[key] = record
    return records


def metric_value(record, metric, discovery_markers):
    if metric in record:
        return record[metric]
    if metric == "clusterCount":
        return len(record["clusters"])
    if metric == "reducedTotal":
        return sum(c["reduced"] for c in record["clusters"])
    if metric in discovery_markers:
        markers = discovery_markers[metric]
        attempts = [
            c["firstSeenAttempt"]
            for c in record["clusters"]
            if all(marker in c["canonicalDescription"] for marker in markers)
        ]
        return min(attempts) if attempts else None
    raise KeyError(f"unknown metric {metric}")


def sign_test(deltas):
    positive = sum(1 for d in deltas if d > 0)
    negative = sum(1 for d in deltas if d < 0)
    n = positive + negative
    if n == 0:
        return positive, negative, 1.0
    k = min(positive, negative)
    tail = sum(math.comb(n, i) for i in range(k + 1)) / 2**n
    return positive, negative, min(1.0, 2 * tail)


def iqr(values):
    ordered = sorted(values)
    def quantile(fraction):
        position = (len(ordered) - 1) * fraction
        low = int(math.floor(position))
        high = int(math.ceil(position))
        return ordered[low] + (ordered[high] - ordered[low]) * (position - low)
    return quantile(0.25), quantile(0.75)


def fmt(value):
    if value is None:
        return "notfound"
    if isinstance(value, float):
        return f"{value:.1f}"
    return str(value)


def analyze_metric(metric, pairs, discovery_markers):
    print(f"\n=== {metric} ===")
    print(f"{'fixture':8} {'seed':>5} {'baseline':>12} {'candidate':>12} {'delta':>12}")
    deltas = []
    baseline_values = []
    candidate_values = []
    for (fixture, seed), (base, cand) in sorted(pairs.items()):
        base_value = metric_value(base, metric, discovery_markers)
        cand_value = metric_value(cand, metric, discovery_markers)
        if base_value is None and cand_value is None:
            print(f"{fixture:8} {seed:>5} {'notfound':>12} {'notfound':>12} {'dropped':>12}")
            continue
        if base_value is None:
            deltas.append(-1)  # candidate found what the baseline never did: an improvement
            print(f"{fixture:8} {seed:>5} {'notfound':>12} {fmt(cand_value):>12} {'improved':>12}")
            continue
        if cand_value is None:
            deltas.append(1)
            print(f"{fixture:8} {seed:>5} {fmt(base_value):>12} {'notfound':>12} {'regressed':>12}")
            continue
        delta = cand_value - base_value
        deltas.append(delta)
        baseline_values.append(base_value)
        candidate_values.append(cand_value)
        print(f"{fixture:8} {seed:>5} {fmt(base_value):>12} {fmt(cand_value):>12} {fmt(delta):>12}")

    if not deltas:
        print("no comparable pairs")
        return
    if baseline_values:
        base_low, base_high = iqr(baseline_values)
        cand_low, cand_high = iqr(candidate_values)
        print(f"baseline : median {fmt(median(baseline_values))}  IQR [{fmt(base_low)}, {fmt(base_high)}]")
        print(f"candidate: median {fmt(median(candidate_values))}  IQR [{fmt(cand_low)}, {fmt(cand_high)}]")
        numeric_deltas = [c - b for b, c in zip(baseline_values, candidate_values)]
        print(f"delta    : median {fmt(median(numeric_deltas))}")
    positive, negative, p_value = sign_test(deltas)
    verdict = "significant at 0.05" if p_value < 0.05 else "not significant"
    print(f"sign test: {positive} up, {negative} down, {len(deltas) - positive - negative} ties dropped; p = {p_value:.4f} ({verdict})")


def main():
    parser = argparse.ArgumentParser(description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)
    parser.add_argument("baseline")
    parser.add_argument("candidate")
    parser.add_argument("--metrics", default=",".join(DEFAULT_METRICS))
    parser.add_argument(
        "--discovery",
        action="append",
        default=[],
        help='label=marker[;marker...] — attempts-to-discovery of the cluster matching all markers',
    )
    arguments = parser.parse_args()

    discovery_markers = {}
    metrics = [m for m in arguments.metrics.split(",") if m]
    for spec in arguments.discovery:
        label, _, markers = spec.partition("=")
        if not markers:
            parser.error(f"--discovery '{spec}' must be label=marker[;marker...]")
        discovery_markers[label] = markers.split(";")
        metrics.append(label)

    baseline = load(arguments.baseline)
    candidate = load(arguments.candidate)
    shared = sorted(set(baseline) & set(candidate))
    missing = sorted(set(baseline) ^ set(candidate))
    if missing:
        print(f"warning: {len(missing)} unpaired records skipped: {missing[:6]}{'...' if len(missing) > 6 else ''}", file=sys.stderr)
    if not shared:
        sys.exit("no paired (fixture, seed) records between the two files")
    pairs = {key: (baseline[key], candidate[key]) for key in shared}
    print(f"{len(pairs)} paired runs: {arguments.baseline} vs {arguments.candidate}")
    for metric in metrics:
        analyze_metric(metric, pairs, discovery_markers)


if __name__ == "__main__":
    main()

File: ExploreHarness/Sources/ExploreBenchmark/Analyze.swift
// Paired A/B analysis of two benchmark JSONL files — the Swift port of the former Benchmarks/analyze.py, kept semantically identical: records pair by (fixture, seed), every comparison is per-seed first and aggregates second, censored discovery runs count as strictly worse than any completed one, and ties drop from the sign test.

import ArgumentParser
import Foundation

extension ExploreBenchmark {
    struct Analyze: ParsableCommand {
        static let configuration = CommandConfiguration(
            commandName: "analyze",
            abstract: "Paired A/B analysis of two benchmark JSONL files.",
            discussion: """
            Records are paired by (fixture, seed); every comparison is per-seed first, aggregates second. For each metric the command prints the per-seed table, then per-fixture medians, IQRs, median deltas, and two-sided paired sign-test verdicts, plus a pooled block when more than one fixture is present. Ties are dropped from the sign test.

            --discovery defines an attempts-to-discovery metric: the minimum firstSeenAttempt over clusters whose canonicalDescription contains every semicolon-separated marker. A run that never found a matching cluster is censored: it counts as strictly worse than any run that did, and censored-in-both pairs are dropped.
            """
        )

        @Argument(help: "Baseline JSONL path.")
        var baseline: String

        @Argument(help: "Candidate JSONL path.")
        var candidate: String

        @Option(help: "Comma-separated metrics. Known: any numeric record field, plus clusterCount and reducedTotal.")
        var metrics: String = "attemptsPerSecond,coveredEdges,clusterCount,reducedTotal"

        @Option(help: "label=marker[;marker...] — attempts-to-discovery of the cluster matching all markers. Repeatable.")
        var discovery: [String] = []

        mutating func run() throws {
            var resolved: [Metric] = []
            for name in metrics.split(separator: ",").map(String.init) {
                guard let metric = Self.recordMetrics[name] else {
                    throw ValidationError("unknown metric '\(name)'. Known metrics: \(Self.recordMetrics.keys.sorted().joined(separator: ", ")).")
                }
                resolved.append(metric)
            }
            for spec in discovery {
                let parts = spec.split(separator: "=", maxSplits: 1, omittingEmptySubsequences: false)
                guard parts.count == 2, parts[1].isEmpty == false else {
                    throw ValidationError("--discovery '\(spec)' must be label=marker[;marker...]")
                }
                let markers = parts[1].split(separator: ";").map(String.init)
                resolved.append(Metric(name: String(parts[0]), kind: .integer) { record in
                    let attempts = record.clusters
                        .filter { cluster in markers.allSatisfy { cluster.canonicalDescription.contains($0) } }
                        .map(\.firstSeenAttempt)
                    return attempts.min().map(Double.init)
                })
            }

            let baselineRecords = try Self.load(baseline)
            let candidateRecords = try Self.load(candidate)
            let shared = Set(baselineRecords.keys).intersection(candidateRecords.keys).sorted()
            let unpaired = Set(baselineRecords.keys).symmetricDifference(candidateRecords.keys).sorted()
            if unpaired.isEmpty == false {
                let preview = unpaired.prefix(6).map { "(\($0.fixture), \($0.seed))" }.joined(separator: ", ")
                warn("warning: \(unpaired.count) unpaired records skipped: \(preview)\(unpaired.count > 6 ? "..." : "")")
            }
            guard shared.isEmpty == false else {
                warn("no paired (fixture, seed) records between the two files")
                throw ExitCode(1)
            }
            print("\(shared.count) paired runs: \(baseline) vs \(candidate)")
            for metric in resolved {
                Self.analyzeMetric(metric, keys: shared, baseline: baselineRecords, candidate: candidateRecords)
            }
        }

        // MARK: - Metrics

        /// The numeric record surface. `handleSkipFraction` reads as censored (`notfound`) on records without it rather than erroring, matching its optional schema.
        static let recordMetrics: [String: Metric] = {
            var table: [String: Metric] = [:]
            func add(_ name: String, _ kind: MetricKind, decimals: Int = 1, _ value: @escaping @Sendable (BenchmarkRecord) -> Double?) {
                table[name] = Metric(name: name, kind: kind, decimals: decimals, value: value)
            }
            add("seed", .integer) { Double($0.seed) }
            add("screeningAttempts", .integer) { Double($0.screeningAttempts) }
            add("samplingAttempts", .integer) { Double($0.samplingAttempts) }
            add("sprawlAttempts", .integer) { Double($0.sprawlAttempts) }
            add("totalAttempts", .integer) { Double($0.totalAttempts) }
            add("discardedAttempts", .integer) { Double($0.discardedAttempts) }
            add("coveredEdges", .integer) { Double($0.coveredEdges) }
            add("instrumentedEdges", .integer) { Double($0.instrumentedEdges) }
            add("edgeSingletons", .integer) { Double($0.edgeSingletons) }
            add("edgeDoubletons", .integer) { Double($0.edgeDoubletons) }
            add("corpusEntries", .integer) { Double($0.corpusEntries) }
            add("clusterCount", .integer) { Double($0.clusters.count) }
            add("reducedTotal", .integer) { Double($0.clusters.reduce(0) { $0 + $1.reduced }) }
            add("budgetSeconds", .fractional) { $0.budgetSeconds }
            add("elapsedSeconds", .fractional) { $0.elapsedSeconds }
            add("attemptsPerSecond", .fractional) { $0.attemptsPerSecond }
            add("overheadFraction", .fractional, decimals: 4) { $0.overheadFraction }
            add("estimatedNextEdgeProbability", .fractional, decimals: 4) { $0.estimatedNextEdgeProbability }
            add("estimatedReachableEdges", .fractional) { $0.estimatedReachableEdges }
            add("handleSkipFraction", .fractional, decimals: 4) { $0.handleSkipFraction }
            return table
        }()

        // MARK: - Loading

        static func load(_ path: String) throws -> [PairKey: BenchmarkRecord] {
            let decoder = JSONDecoder()
            var records: [PairKey: BenchmarkRecord] = [:]
            let content = try String(contentsOfFile: path, encoding: .utf8)
            for rawLine in content.split(separator: "\n", omittingEmptySubsequences: true) {
                let line = rawLine.trimmingCharacters(in: .whitespaces)
                if line.isEmpty {
                    continue
                }
                let record = try decoder.decode(BenchmarkRecord.self, from: Data(line.utf8))
                let key = PairKey(fixture: record.fixture, seed: record.seed)
                if records[key] != nil {
                    warn("warning: duplicate record for (\(key.fixture), \(key.seed)) in \(path); keeping the last one")
                }
                records[key] = record
            }
            return records
        }

        // MARK: - Analysis

        static func analyzeMetric(
            _ metric: Metric,
            keys: [PairKey],
            baseline: [PairKey: BenchmarkRecord],
            candidate: [PairKey: BenchmarkRecord]
        ) {
            print("\n=== \(metric.name) ===")
            print("\(pad("fixture", 18)) \(padLeft("seed", 5)) \(padLeft("baseline", 12)) \(padLeft("candidate", 12)) \(padLeft("delta", 12))")
            var byFixture: [String: Accumulator] = [:]
            var fixtureOrder: [String] = []
            for key in keys {
                guard let baseRecord = baseline[key], let candidateRecord = candidate[key] else {
                    continue
                }
                if byFixture[key.fixture] == nil {
                    byFixture[key.fixture] = Accumulator()
                    fixtureOrder.append(key.fixture)
                }
                let baseValue = metric.value(baseRecord)
                let candidateValue = metric.value(candidateRecord)
                let row = { (baseText: String, candidateText: String, deltaText: String) in
                    print("\(pad(key.fixture, 18)) \(padLeft("\(key.seed)", 5)) \(padLeft(baseText, 12)) \(padLeft(candidateText, 12)) \(padLeft(deltaText, 12))")
                }
                switch (baseValue, candidateValue) {
                    case (nil, nil):
                        row("notfound", "notfound", "dropped")
                    case let (nil, .some(found)):
                        // The candidate found what the baseline never did: an improvement for the sign test.
                        byFixture[key.fixture]?.deltas.append(-1)
                        row("notfound", fmt(found, metric), "improved")
                    case let (.some(found), nil):
                        byFixture[key.fixture]?.deltas.append(1)
                        row(fmt(found, metric), "notfound", "regressed")
                    case let (.some(base), .some(cand)):
                        byFixture[key.fixture]?.deltas.append(cand - base)
                        byFixture[key.fixture]?.baselineValues.append(base)
                        byFixture[key.fixture]?.candidateValues.append(cand)
                        row(fmt(base, metric), fmt(cand, metric), fmt(cand - base, metric))
                }
            }

            // Gates read per-fixture verdicts (per-class no-regression); the pooled block stays for single-fixture runs and quick overall reads.
            for fixture in fixtureOrder.sorted() {
                guard let accumulator = byFixture[fixture] else {
                    continue
                }
                summarize("[\(fixture)]", accumulator, metric: metric)
            }
            if fixtureOrder.count > 1 {
                var pooled = Accumulator()
                for accumulator in byFixture.values {
                    pooled.deltas.append(contentsOf: accumulator.deltas)
                    pooled.baselineValues.append(contentsOf: accumulator.baselineValues)
                    pooled.candidateValues.append(contentsOf: accumulator.candidateValues)
                }
                summarize("[all fixtures]", pooled, metric: metric)
            }
        }

        private static func summarize(_ label: String, _ accumulator: Accumulator, metric: Metric) {
            guard accumulator.deltas.isEmpty == false else {
                print("\(label): no comparable pairs")
                return
            }
            print(label)
            if accumulator.baselineValues.isEmpty == false {
                let (baseLow, baseHigh) = iqr(accumulator.baselineValues)
                let (candidateLow, candidateHigh) = iqr(accumulator.candidateValues)
                print("  baseline : median \(fmtMedian(accumulator.baselineValues, metric))  IQR [\(fmtFractional(baseLow, metric.decimals)), \(fmtFractional(baseHigh, metric.decimals))]")
                print("  candidate: median \(fmtMedian(accumulator.candidateValues, metric))  IQR [\(fmtFractional(candidateLow, metric.decimals)), \(fmtFractional(candidateHigh, metric.decimals))]")
                let numericDeltas = zip(accumulator.baselineValues, accumulator.candidateValues).map { $1 - $0 }
                print("  delta    : median \(fmtMedian(numericDeltas, metric))")
            }
            let (positive, negative, pValue) = signTest(accumulator.deltas)
            let ties = accumulator.deltas.count - positive - negative
            let verdict = pValue < 0.05 ? "significant at 0.05" : "not significant"
            print("  sign test: \(positive) up, \(negative) down, \(ties) ties dropped; p = \(String(format: "%.4f", pValue)) (\(verdict))")
        }

        // MARK: - Statistics

        /// Two-sided paired sign test at p = 1/2, ties already excluded by construction. The binomial tail accumulates iteratively in probability space; the smallest term, 0.5^n, stays representable far past any realistic pair count.
        static func signTest(_ deltas: [Double]) -> (positive: Int, negative: Int, pValue: Double) {
            let positive = deltas.count(where: { $0 > 0 })
            let negative = deltas.count(where: { $0 < 0 })
            let n = positive + negative
            if n == 0 {
                return (positive, negative, 1.0)
            }
            let k = min(positive, negative)
            var probabilityMass = pow(0.5, Double(n))
            var tail = probabilityMass
            for i in 0 ..< k {
                probabilityMass = probabilityMass * Double(n - i) / Double(i + 1)
                tail += probabilityMass
            }
            return (positive, negative, min(1.0, 2 * tail))
        }

        /// The true median: the middle element, or the mean of the two middles for even counts.
        static func median(_ values: [Double]) -> Double {
            let ordered = values.sorted()
            let count = ordered.count
            if count % 2 == 1 {
                return ordered[count / 2]
            }
            return (ordered[count / 2 - 1] + ordered[count / 2]) / 2
        }

        /// Linearly interpolated quartiles, matching the former script's quantile convention.
        static func iqr(_ values: [Double]) -> (low: Double, high: Double) {
            let ordered = values.sorted()
            func quantile(_ fraction: Double) -> Double {
                let position = Double(ordered.count - 1) * fraction
                let low = Int(position.rounded(.down))
                let high = Int(position.rounded(.up))
                return ordered[low] + (ordered[high] - ordered[low]) * (position - Double(low))
            }
            return (low: quantile(0.25), high: quantile(0.75))
        }

        // MARK: - Formatting

        /// Integer metrics print whole values plainly and averaged values (medians of even counts) with one decimal; fractional metrics print at the metric's declared precision.
        private static func fmt(_ value: Double, _ metric: Metric) -> String {
            if metric.kind == .integer, value == value.rounded() {
                return String(Int(value))
            }
            return String(format: "%.\(metric.decimals)f", value)
        }

        /// Interpolated quantiles always print fractionally, at the metric's declared precision.
        private static func fmtFractional(_ value: Double, _ decimals: Int = 1) -> String {
            String(format: "%.\(decimals)f", value)
        }

        /// An even-count median averages the two middles and therefore prints fractionally even for integer metrics; an odd-count median is an element and follows the metric kind.
        private static func fmtMedian(_ values: [Double], _ metric: Metric) -> String {
            values.count % 2 == 0 ? fmtFractional(median(values), metric.decimals) : fmt(median(values), metric)
        }

        private static func pad(_ text: String, _ width: Int) -> String {
            text.count >= width ? text : text + String(repeating: " ", count: width - text.count)
        }

        private static func padLeft(_ text: String, _ width: Int) -> String {
            text.count >= width ? text : String(repeating: " ", count: width - text.count) + text
        }
    }

    // MARK: - Analysis Types

    enum MetricKind {
        case integer
        case fractional
    }

    struct Metric: Sendable {
        let name: String
        let kind: MetricKind
        /// Display decimals for fractional values. Fraction-scale observables (skip fraction, overhead) need more than the default one, or a 0.025-scale signal prints as 0.0.
        var decimals: Int = 1
        let value: @Sendable (BenchmarkRecord) -> Double?
    }

    struct PairKey: Hashable, Comparable {
        let fixture: String
        let seed: UInt64

        static func < (lhs: PairKey, rhs: PairKey) -> Bool {
            lhs.fixture == rhs.fixture ? lhs.seed < rhs.seed : lhs.fixture < rhs.fixture
        }
    }

    /// Per-fixture accumulation for one metric: sign-test deltas (including the ±1 censoring sentinels) and the numeric value pairs.
    private struct Accumulator {
        var deltas: [Double] = []
        var baselineValues: [Double] = []
        var candidateValues: [Double] = []
    }
}

private func warn(_ message: String) {
    FileHandle.standardError.write(Data((message + "\n").utf8))
}
